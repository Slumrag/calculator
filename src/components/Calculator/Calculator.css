.Calculator {
  max-width: 554px;
  aspect-ratio: 0.692;
  padding: 10px 44px 46px;
  display: flex;
  flex-direction: column;

  background: linear-gradient(155.233deg, #28518e 0%, #3a77d1 100%);
  border-radius: 18px;

  box-shadow: 0px 82px 158px 0px rgba(0, 0, 0, 35%), 0px 24px 47px 0px rgba(0, 0, 0, 23%),
    0px 10px 19px 0px rgba(0, 0, 0, 17%), 0px 3px 7px 0px rgba(0, 0, 0, 12%);
}

.Calculator__history {
  flex: 1;
  font-size: 24px;
  line-height: 32px;
  display: flex;
  align-items: end;
  justify-content: end;
}

.Calculator__input {
  margin-bottom: 18px;
  margin-top: 28px;

  font-size: 56px;
  font-weight: 800;
  text-align: end;
  line-height: 80px;
}

.Calculator__divider {
  width: 100%;
  height: 2px;
  margin-top: -2px;
  background: rgba(255, 255, 255, 0.35);
}

.Calculator__numpad {
  margin-top: 37px;
  align-self: center;
  display: grid;
  grid-template-rows: repeat(5, 80px);
  grid-template-columns: repeat(4, 100px);
  gap: 16px 20px;
  place-items: center;
}

.Calculator__numpad-button {
  width: 80px;
  height: 80px;
  font-size: 36px;
  border-radius: 50%;
  transition: background 0.2s ease-in-out;
}

.Calculator__numpad-button_solid {
  background: #ffffff;
  color: #2b589a;
}

@media (hover: hover) {
  .Calculator__numpad-button:hover {
    background-color: rgba(255, 255, 255, 12%);
  }
  .Calculator__numpad-button_solid:hover {
    background-color: rgba(255, 255, 255, 88%);
  }
}

@media (hover: none) {
  .Calculator__numpad-button:active {
    background-color: rgba(255, 255, 255, 12%);
  }
  .Calculator__numpad-button_solid:active {
    background-color: rgba(255, 255, 255, 88%);
  }
}
