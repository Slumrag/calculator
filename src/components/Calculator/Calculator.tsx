import './Calculator.css';

function Calculator() {
  return (
    <div className='Calculator'>
      <div className='Calculator__body'>
        <p className='Calculator__history'>20×80×0,5</p>
        <input className='Calculator__input' type='numeric' value={`500`} />
        <div className='Calculator__divider'></div>
        <div className='Calculator__numpad'>
          <button className='Calculator__numpad-button' id='clear'>
            C
          </button>
          <button className='Calculator__numpad-button' id='sqrt'>
            √
          </button>
          <button className='Calculator__numpad-button' id='percent'>
            %
          </button>
          <button className='Calculator__numpad-button' id='divide'>
            /
          </button>
          {/* row1 */}
          <button className='Calculator__numpad-button' id='7'>
            7
          </button>
          <button className='Calculator__numpad-button' id='8'>
            8
          </button>

          <button className='Calculator__numpad-button' id='9'>
            9
          </button>
          <button className='Calculator__numpad-button' id='multiply'>
            ×
          </button>
          {/* row2 */}
          <button className='Calculator__numpad-button' id='4'>
            4
          </button>

          <button className='Calculator__numpad-button' id='5'>
            5
          </button>
          <button className='Calculator__numpad-button' id='6'>
            6
          </button>
          <button className='Calculator__numpad-button' id='subtract'>
            -
          </button>
          {/* row3 */}
          <button className='Calculator__numpad-button' id='1'>
            1
          </button>

          <button className='Calculator__numpad-button' id='2'>
            2
          </button>
          <button className='Calculator__numpad-button' id='3'>
            3
          </button>
          <button className='Calculator__numpad-button' id='add'>
            +
          </button>

          {/* row4 */}
          <button className='Calculator__numpad-button' id='00'>
            00
          </button>
          <button className='Calculator__numpad-button' id='0'>
            0
          </button>
          <button className='Calculator__numpad-button' id=','>
            ,
          </button>
          <button className='Calculator__numpad-button Calculator__numpad-button_solid' id='equals'>
            =
          </button>
        </div>
      </div>
    </div>
  );
}

export default Calculator;
