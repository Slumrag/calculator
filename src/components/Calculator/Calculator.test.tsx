import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Calculator from './Calculator';

describe('Calculator', () => {
  it('should match snapshot', () => {
    render(<Calculator />);
    expect(screen.getByTestId('Calculator')).toMatchSnapshot();
  });

  describe('numpad interactions', () => {
    it('should input correct values on numpad click', async () => {
      const user = userEvent.setup();
      render(<Calculator />);
      const input: HTMLInputElement = screen.getByTestId('Calculator__input');
      const numpad = screen.getAllByRole('button');
      screen.debug(numpad);
      await user.click(numpad[4]);
      expect(input.value).toBe('7');
    });

    it('should clear input on clicking C', async () => {
      const user = userEvent.setup();
      render(<Calculator />);
      const input: HTMLInputElement = screen.getByTestId('Calculator__input');
      await user.type(input, '123');
      // screen.debug(numpad);
      await user.click(screen.getByTestId('clear'));
      expect(input.value).toBe('');
    });

    it.todo('should evaluate expression clicking =');
  });

  describe('input interactions', () => {
    it.todo('should clear input on pressing Escape');
    it.todo('should evaluate input on pressing Enter');
  });
  describe('history interaction', () => {
    it.todo('should display equation after evaluation');
  });
});
