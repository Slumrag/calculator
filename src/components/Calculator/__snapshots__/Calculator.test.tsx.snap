// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calculator > should match snapshot 1`] = `
<div
  class="Calculator"
  data-testid="Calculator"
>
  <div
    class="Calculator__body"
  >
    <p
      class="Calculator__history"
      data-testid="Calculator__history"
    />
    <input
      class="Calculator__input"
      data-testid="Calculator__input"
      type="numeric"
      value=""
    />
    <div
      class="Calculator__divider"
    />
    <div
      class="Calculator__numpad"
      data-testid="Calculator__numpad"
    >
      <button
        class="Calculator__numpad-button"
        data-testid="clear"
        id="clear"
      >
        C
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="sqrt"
        id="sqrt"
      >
        √
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="percent"
        id="percent"
      >
        %
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="divide"
        id="divide"
      >
        /
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="7"
        id="7"
      >
        7
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="8"
        id="8"
      >
        8
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="9"
        id="9"
      >
        9
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="multiply"
        id="multiply"
      >
        ×
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="4"
        id="4"
      >
        4
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="5"
        id="5"
      >
        5
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="6"
        id="6"
      >
        6
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="subtract"
        id="subtract"
      >
        -
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="1"
        id="1"
      >
        1
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="2"
        id="2"
      >
        2
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="3"
        id="3"
      >
        3
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="add"
        id="add"
      >
        +
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="00"
        id="00"
      >
        00
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid="0"
        id="0"
      >
        0
      </button>
      <button
        class="Calculator__numpad-button"
        data-testid=","
        id=","
      >
        ,
      </button>
      <button
        class="Calculator__numpad-button Calculator__numpad-button_solid"
        data-testid="equals"
        id="equals"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
